cpp/src/arrow/compute/kernels/vector_rank.cc

```diff
@@ -28,114 +28,97 @@ namespace {
 // ----------------------------------------------------------------------
 // Rank implementation
 
-template <typename ValueSelector,
-          typename T = std::decay_t<std::invoke_result_t<ValueSelector, int64_t>>>
+// A bit that is set in the sort indices when the value at the current sort index
+// is the same as the value at the previous sort index.
+constexpr uint64_t kDuplicateMask = 1ULL << 63;
+
+constexpr bool NeedsDuplicates(RankOptions::Tiebreaker tiebreaker) {
+  return tiebreaker != RankOptions::First;
+}
+
+template <typename ValueSelector>
+void MarkDuplicates(const NullPartitionResult& sorted, ValueSelector&& value_selector) {
+  using T = decltype(value_selector(int64_t{}));
+
+  // Process non-nulls
+  if (sorted.non_nulls_end != sorted.non_nulls_begin) {
+    auto it = sorted.non_nulls_begin;
+    T prev_value = value_selector(*it);
+    while (++it < sorted.non_nulls_end) {
+      T curr_value = value_selector(*it);
+      if (curr_value == prev_value) {
+        *it |= kDuplicateMask;
```

Comment:
Shall we somehow expose the index type of `GenericNullPartitionResult` and `static_assert` somewhere in this function that it is `uint64_t`?

Reply:
I'm not sure the `static_assert` would even be necessary, as long as `value_selector` can take the index type as an argument. I'll give it a try.

Reply:
It worked indeed.

Reply:
Sorry I didn't see your previous minor change. Did you mean that the index type being `uint64_t` is already guaranteed by `*it |= kDuplicateMask` so that we don't need any `static_assert`?

Reply:
The "minor change" was to make `MarkDuplicates` generic on the index type. But then I realize it couldn't, since `*it |= kDuplicateMask` requires the index to be a 64-bit-int. So I reverted to the specific version.

Reply:
I'm OK with it. Thank you.